shpc/main/modules/base.py

```diff
@@ -480,3 +497,49 @@ def install(
             view.install(module_dir)
 
         return container_path
+
+    def reinstall_all(
+        self, pattern=False, view=None, disable_view=False, force=False, **kwargs
+    ):
+        """
+        Reinstall (and possibly upgrade) all the current modules, possibly filtered by pattern.
+        """
+        modules = self._get_module_lookup(
+            self.settings.module_base, self.modulefile, pattern
+        )
+
+        # If we don't have modules, exit early
+        if not modules:
+            logger.exit("You don't have any install modules. Try shpc show.", 0)
+
+        unavailable_modules = False
```

Comment:
Ok this is getting a bit hairy - my instinct is saying we want to break this down into a few steps.

1. get a list of installed modules and all modules. We can derive those already installed vs. not installed yet with basic sets, and OK to add a version to that if desired.
2. I think we probably want to just call the self.install function directly from self.reinstall
3. I think we should assume a reinstall all is going for the latest version - no need to check and spit out a warning. The one warning we want (and a continue in the loop) is if something is installed but we don't have a source anymore. In this case it's just ignored (not deleted)
4. And there might be a `def reinstall` function to handle operations for a single case, with optional arguments for the lookup of already installed vs. known, and then someone could call it one off for one package:

```bash
$ shpc reinstall python
This will reinstall python@3.5.1 as python@3.9.1 are you sure [Y/N]?
```

And then the reinstall all case:

```bash
$ shpc reinstall --all
```
And they would be using the same underlying functions. And even for the large "all" case we might want to do this interactively, where we essentially create a table of installed vs will be installed (with empty spaces where appropriate) and then ask the user (as we do above) y/n "Do you want to continue?" If no, they might have the choice to specify a specific version pin, e.g.,

```bash
$ shpc reinstall --all --pin python@3.5.1
$ shpc reinstall --all -p python@3.5.1
```
If you are overwhelmed (or just short of time) I can probably take a crack this weekend! I think this feature will be great to have! Let me know if you want more detail or pseudocode for any of the above!

Reply:
Thanks for the feedback. Maybe the name `reinstall` isn't quite right. The use case I've got in mind is to regenerate all modules & co when shpc itself has a new feature, e.g. wrapper scripts for aliases, wrapper scripts for helpers, etc, while leaving all the versions as they are. I would disagree with "we should assume a reinstall all is going for the latest version". But I agree a general upgrade (like `brew upgrade`) would be nice to add.
Would you be happy making the argument of `--pin` optional ? therefore allowing `shpc reinstall --all --pin` to fulfil the scenario I've got in mind. I actually already had a commit that adds some CLI parameters for reinstall, and I didn't push it because I wanted to think more about the CLI interface, e.g. the `shpc reinstall python` you're proposing would do the same as `shpc install python` and that'd confuse me as a user.

I'm unlikely to have any time this week to implement this, but I can still discuss the feature

Reply:
Ah I think I better understand now!

> Would you be happy making the argument of --pin optional ?

Yes absolutely - and if what I'm suggesting would just emulate `shpc upgrade --all` then perhaps we don't need `--pin` (and the behavior you describe would be the default.

> shpc reinstall python you're proposing would do the same as shpc install python and that'd confuse me as a user.

The difference is that `shpc install` would get latest, and `shpc reinstall` would try to re-create the existing version.